subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
 			"order may change at reboot.");
 	}
 
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT) && ak_overwritten) {
+		/* Account Key overwritten. Remove bonds releted with overwritten Account Key. */
+		bond_remove_old_id();
+	}
+
 	return 0;
 }
 
+static int ak_metadata_id_get(uint8_t *id, const struct fp_account_key *account_key)
+{
+	for (size_t i = 0; i < account_key_count; i++) {
+		if (!memcmp(account_key_list[i].key, account_key->key, FP_ACCOUNT_KEY_LEN)) {
+			*id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
+			return 0;
+		}
+	}
+
+	return -ESRCH;
+}
+
+static int bond_free_idx_get(size_t *bond_idx)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			*bond_idx = i;
+			return 0;
+		}
+	}
+
+	return -ENOMEM;
+}
+
+static int bonded_duplicate_find(size_t *bond_idx, const bt_addr_le_t *addr, const void *conn_ctx)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if ((fp_bonds[i].conn_ctx != conn_ctx) &&
+		    fp_bonds[i].bonded &&
+		    bt_addr_le_eq(&fp_bonds[i].nv.addr, addr)) {
+			*bond_idx = i;
+			return 0;
+		}
+	}
+
+	return -ESRCH;
+}
+
+static void bond_update_addr(const bt_addr_le_t *new_addr, const void *conn_ctx)
+{
+	int err;
+	size_t bond_idx;
+	struct fp_bond bond;
+	char name[SETTINGS_BOND_NAME_MAX_SIZE];
+
+	err = bond_find(&bond_idx, conn_ctx);
+	if (err) {
+		return;
+	}
+
+	bond = fp_bonds[bond_idx];
+
+	bond.nv.addr = *new_addr;
+
+	err = bond_name_gen(name, bond_idx);
+	if (err) {
+		LOG_ERR("Failed to generate bond name (err %d)", err);
+		return;
+	}
+
+	err = settings_save_one(name, &bond.nv, sizeof(bond.nv));
+	if (err) {
+		LOG_ERR("Failed to save bond (err %d)", err);
+		return;
+	}
+
+	fp_bonds[bond_idx] = bond;
+}
+
+static int bonded_by_addr_find(size_t *bond_idx, const bt_addr_le_t *addr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (fp_bonds[i].bonded &&
+		    bt_addr_le_eq(&fp_bonds[i].nv.addr, addr)) {
+			*bond_idx = i;
+			return 0;
+		}
+	}
+
+	return -ESRCH;
+}
+
+void fp_storage_ak_bond_bt_request_cb_register(const struct fp_storage_ak_bond_bt_request_cb *cb)
+{
+	__ASSERT_NO_MSG(IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT));
+	__ASSERT_NO_MSG(cb);
+
+	bt_request_cb = cb;
+}
+
+int fp_storage_ak_bond_save(const bt_addr_le_t *addr, const void *conn_ctx,
+			    const struct fp_account_key *account_key)
+{
+	__ASSERT_NO_MSG(IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT));
+
+	if (!is_enabled) {
+		return -EACCES;
+	}
+
+	int err;
+	uint8_t id;
+	size_t bond_idx;
+	struct fp_bond bond;
+	char name[SETTINGS_BOND_NAME_MAX_SIZE];
+
+	if (account_key) {
+		err = ak_metadata_id_get(&id, account_key);
+		if (err) {
+			return err;
+		}
+	} else {
+		/* The actual ID will be determined during Account Key write. */
+		id = 0;
+	}
+
+	err = bond_free_idx_get(&bond_idx);
```

Comment:
Maybe you could return pointer to found `struct fp_bond` and fill it directly to simplify implementation and avoid copying? I think you could also safely assert on error here

Reply:
You could event make something like `fp_bond_find(conn_ctx)` where `conn_ctx == NULL` would be used to find free structure